### PR TITLE
#165305772 Users can bookmark an article

### PR DESCRIPTION
**Pull request by Amoswachira**
### What does this PR do?
- Allows users to bookmark an article
### What are the tasks to be completed?
```
GET  /api/article/bookmarks/
POST /api/articles/{slug}/bookmark/
DELETE /api/articles/bookmark/{slug}/
```
* Ensure that a user can be able to bookmark an article
* Ensure a user can be able to delete a bookmark
* Ensure a user can be able to get his bookmarks

### How can this be manually tested?
 **After cloning the repo:-**
**Test with postman**
Sign up a user, login, grub the token and add it to authorization:
```
$ git checkout  ft-bookmark-article-165305772 
$ python manage.py runserver
```
**To run tests on terminal**
```
$ python manage.py test authors/apps/articles
```

### What are the relevant pivotal tracker stories?
[#165305772](https://www.pivotaltracker.com/n/projects/2326460/stories/165305772)

### Screenshots
Screenshot 2019-05-17 at 13 35 00
Screenshot 2019-05-17 at 13 56 11
Screenshot 2019-05-17 at 16 16 31

### CODE REVIEW

File: authors/apps/articles/views.py [hunk 1]
```diff
@@ -591,3 +593,71 @@ def delete(self, request, **kwargs):
                 raise Forbidden
         except Comment.DoesNotExist:
             raise ItemDoesNotExist
+
+
+class BookmarkAPIView(GenericAPIView):
+    """
+    class for Bookmarking and unbookmarking
+    """
+    permission_classes = (IsAuthenticated,)
```
Comment (Bakley) [inline authors/apps/articles/views.py:709]: @Amoswachira You should also add allow `readonly` for the purpose of the `GET` request

Comment (Amoswachira) [reply authors/apps/articles/views.py:709]: @Bakley This is a protected route only Authenticated users can get their bookmarks

**Review by Ogutu-Brian — CHANGES_REQUESTED** [review]
Hello @Amoswachira ,
After looking at how the urls are defined below: I decided to make a delete request with a url that it should not support and this led to the error below. Please handle the error so that our app remains stable all times.
image
image
**The error was triggered by this url patten**
image
Since you used the same class for 3 different patterns, delete request could still be exposed by this url pattern.

**The same patterns affect GET all bookmarks endpoint too**
image

**Review by Ogutu-Brian — CHANGES_REQUESTED** [review]
Hello @Amoswachira ,
Since what the client is looking for are bookmarks but not articles, you could improve on this message to be about bookmarks e.g `You have not bookmarked any article`
image

**Comment by oluwagbenga-joloko** [comment]
I think you should return the articles a user has bookmarked, rather than the bookmark dictionary when getting a list of bookmarked articles. you should also add a field to the articles that show the bookmark status. eg `bookmarked: True`

File: authors/apps/articles/views.py (hunk 1 above)
Comment (IssaIan) [reply authors/apps/articles/views.py:709]: Yes. I agree with @Amoswachira. Only authenticated users should get their bookmarks.

**Comment by Amoswachira** [comment]
@oluwagbenga-joloko  and @Ogutu-Brian  let me work on that right away

**Comment by Amoswachira** [comment]
@Ogutu-Brian  I have resolved your feedback  on the urls issue

**Comment by Amoswachira** [comment]
@oluwagbenga-joloko  I have fixed the issue

**Review by Bakley — APPROVED** [review]

**Review by Ogutu-Brian — APPROVED** [review]

**Review by IssaIan — APPROVED** [review]

**Pull request merged**